src/set.rs

```diff
@@ -0,0 +1,1114 @@
+    ///
+    /// Like `Vec::swap_remove`, the value is removed by swapping it with the
+    /// last element of the set and popping it off. **This perturbs
+    /// the postion of what used to be the last element!**
+    ///
+    /// Return `None` if `value` was not in the set.
+    pub fn swap_remove_full<Q: ?Sized>(&mut self, value: &Q) -> Option<(usize, T)>
+        where Q: Hash + Equivalent<T>,
+    {
+        self.map.swap_remove_full(value).map(|(i, x, ())| (i, x))
+    }
+
+    /// Remove the last value
+    ///
+    /// Computes in **O(1)** time (average).
+    pub fn pop(&mut self) -> Option<T> {
+        self.map.pop().map(|(x, ())| x)
+    }
+
+    /// Scan through each value in the set and keep those where the
+    /// closure `keep` returns `true`.
+    ///
+    /// The order the elements are visited is not specified.
+    ///
+    /// Computes in **O(n)** time (average).
+    pub fn retain<F>(&mut self, mut keep: F)
+        where F: FnMut(&T) -> bool,
+    {
+        self.map.retain(move |x, &mut ()| keep(x))
+    }
+
+    /// Sort the values of the set and return a by value iterator of
+    /// the values with the result.
+    ///
+    /// The sort is stable.
+    pub fn sorted_by<F>(self, mut cmp: F) -> IntoIter<T>
+        where F: FnMut(&T, &T) -> Ordering
+    {
+        IntoIter {
+            iter: self.map.sorted_by(move |a, &(), b, &()| cmp(a, b)),
+        }
+    }
+
+    /// Clears the `OrderSet`, returning all values as a drain iterator.
+    /// Keeps the allocated memory for reuse.
+    pub fn drain(&mut self, range: RangeFull) -> Drain<T> {
+        Drain {
+            iter: self.map.drain(range),
+        }
+    }
+}
+
+impl<T, S> OrderSet<T, S> {
+    /// Get a value by index
+    ///
+    /// Valid indices are *0 <= index < self.len()*
+    ///
+    /// Computes in **O(1)** time.
+    pub fn get_index(&self, index: usize) -> Option<&T> {
+        self.map.get_index(index).map(|(x, &())| x)
+    }
+
+    /// Remove the key-value pair by index
+    ///
+    /// Valid indices are *0 <= index < self.len()*
+    ///
+    /// Computes in **O(1)** time (average).
+    pub fn swap_remove_index(&mut self, index: usize) -> Option<T> {
+        self.map.swap_remove_index(index).map(|(x, ())| x)
+    }
+}
+
+
+pub struct IntoIter<T> {
+    iter: super::IntoIter<T, ()>,
+}
+
+impl<T> Iterator for IntoIter<T> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next().map(|(x, ())| x)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.iter.size_hint()
+    }
+
+    fn count(self) -> usize {
+        self.iter.len()
+    }
+
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        self.iter.nth(n).map(|(x, ())| x)
+    }
+
+    fn last(mut self) -> Option<Self::Item> {
+        self.next_back()
+    }
+
+    fn collect<C>(self) -> C
+        where C: FromIterator<Self::Item>
+    {
+        self.iter.map(|(x, ())| x).collect()
```

Comment:
For this to reach its goal -- delegate down to the sliceiterator.map(f).collect() -- we need to not go through the OrderMap iterator here, right? The reason we delegate down to the libcore iterators is to get their TrustedLen etc optimizations for collect.

Reply:
The other iterator methods should be free of this problem.

Reply:
Ah, I didn't really understand *why* you were overriding `collect()` at all, but it makes sense to leverage the unstable traits this way.  I can make this use `self.iter.iter.map(...).collect()` -- although once we're digging into private fields, maybe it would be better for `set` to just create and use the underlying vec/slice iterator directly, avoiding the doubled `map(..)` in the other methods.

I was *mostly* trying to avoid having the set use private features of the map, but maybe that's not so helpful.  I was somewhat planning for the possibility that you might rather publish this as a separate `orderset` crate.

Reply:
I think it is fine to ship in the same crate, they are closely related.

Reply:
Done.